### PR TITLE
[SPARK-8498] [SQL] Add regression test for SPARK-8470

### PR DESCRIPTION
**Pull request by andrewor14**
**Summary of the problem in SPARK-8470.** When using `HiveContext` to create a data frame of a user case class, Spark throws `scala.reflect.internal.MissingRequirementError` when it tries to infer the schema using reflection. This is caused by `HiveContext` silently overwriting the context class loader containing the user classes.

**What this issue is about.** This issue adds regression tests for SPARK-8470, which is already fixed in #6891. We closed SPARK-8470 as a duplicate because it is a different manifestation of the same problem in SPARK-8368. Due to the complexity of the reproduction, this requires us to pre-package a special test jar and include it in the Spark project itself.

I tested this with and without the fix in #6891 and verified that it passes only if the fix is present.

### CODE REVIEW

**Comment by andrewor14** [comment]
@marmbrus @yhuai

**Comment by andrewor14** [comment]
@marmbrus @yhuai

**Comment by andrewor14** [comment]
retest this please

**Comment by yhuai** [comment]
Merging it to master and branch 1.4.

File: sql/hive/src/test/resources/regression-test-SPARK-8498/Main.scala [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import org.apache.spark.{SparkConf, SparkContext}
+import org.apache.spark.sql.hive.HiveContext
+
+/**
+ * Entry point in test application for SPARK-8498.
```
Comment (andrewor14) [inline sql/hive/src/test/resources/regression-test-SPARK-8498/Main.scala]: Note that this is supposed to be SPARK-8489. This is corrected in a future commit.

**Pull request closed without merging**